### PR TITLE
add observability topology for PowerScale, UT, e2e and etc

### PR DESCRIPTION
**Pull request by YianZong**
# Description
add observability topology for PowerScale, UT, e2e and etc

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/488 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] UT
Current coverage:
`[root@vpi6179 controllers]# go test ./... -coverprofile cover.out
ok      github.com/dell/csm-operator/controllers        0.591s  coverage: 88.5% of statements
[root@vpi6179 modules]# go test ./... -coverprofile cover.out
ok      github.com/dell/csm-operator/pkg/modules        0.105s  coverage: 88.9% of statements
`
- [x] E2E
`Running Suite: CSM Operator End-to-End Tests
============================================
Random Seed: 1664430799 - Will randomize all specs
Will run 1 of 1 specs

STEP: Getting test environment variables
STEP: Reading values file
STEP: Getting a k8s client
[run-e2e-test]E2E Testing
  Running all test Given Test Scenarios
  /home/dellemc/csm-operator/tests/e2e/e2e_test.go:75
[It] Running all test Given Test Scenarios
  /home/dellemc/csm-operator/tests/e2e/e2e_test.go:75
STEP: Starting: Install PowerScale Driver(With Observability)
STEP:      Executing  Given an environment with k8s or openshift, and CSM operator installed
STEP:      Executing  Apply custom resources
Sep 29 01:53:23.261: INFO: Running '/usr/bin/kubectl --namespace=test-isilon apply --validate=true -f -'
Sep 29 01:53:23.507: INFO: stderr: ""
Sep 29 01:53:23.507: INFO: stdout: "containerstoragemodule.storage.dell.com/test-isilon created\n"
STEP:      Executing  Validate custom resources
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is installed
STEP:      Executing  Enable forceRemoveDriver on CR
STEP:      Executing  Delete resources
STEP: Ending: Install PowerScale Driver(With Observability)
STEP:
STEP:
STEP: Starting: Install PowerScale Driver(Standalone), Enable Observability
STEP:      Executing  Given an environment with k8s or openshift, and CSM operator installed
STEP:      Executing  Apply custom resources
Sep 29 01:53:38.553: INFO: Running '/usr/bin/kubectl --namespace=test-isilon apply --validate=true -f -'
Sep 29 01:53:38.802: INFO: stderr: ""
Sep 29 01:53:38.802: INFO: stdout: "containerstoragemodule.storage.dell.com/test-isilon created\n"
STEP:      Executing  Validate custom resources
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is not installed
STEP:      Executing  Enable [observability] module
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is installed
STEP:      Executing  Enable forceRemoveDriver on CR
STEP:      Executing  Delete resources
STEP: Ending: Install PowerScale Driver(Standalone), Enable Observability
STEP:
STEP:
STEP: Starting: Install PowerScale Driver(With Observability), Disable Observability module"
STEP:      Executing  Given an environment with k8s or openshift, and CSM operator installed
STEP:      Executing  Apply custom resources
Sep 29 01:53:53.878: INFO: Running '/usr/bin/kubectl --namespace=test-isilon apply --validate=true -f -'
Sep 29 01:53:54.166: INFO: stderr: ""
Sep 29 01:53:54.166: INFO: stdout: "containerstoragemodule.storage.dell.com/test-isilon created\n"
STEP:      Executing  Validate custom resources
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is installed
STEP:      Executing  Disable [observability] module
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is not installed
STEP:      Executing  Enable forceRemoveDriver on CR
STEP:      Executing  Delete resources
STEP: Ending: Install PowerScale Driver(With Observability), Disable Observability module"
STEP:
STEP:

• [SLOW TEST:55.986 seconds]
[run-e2e-test]E2E Testing
/home/dellemc/csm-operator/tests/e2e/e2e_test.go:74
  Running all test Given Test Scenarios
  /home/dellemc/csm-operator/tests/e2e/e2e_test.go:75
------------------------------

Ran 1 of 1 Specs in 56.344 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 59.587013979s
Test Suite Passed

`
- [x] Manual
image

### CODE REVIEW

**Comment by rajendraindukuri** [comment]
Hi @YianZong Please do not merge this PR as posting activities are in progress for CSM Operator. Once those are done, we can merge. Blocking this PR for now.

File: operatorconfig/moduleconfig/common/version-values.yaml
Comment (forrestxia) [inline operatorconfig/moduleconfig/common/version-values.yaml:13]: different obs modules have different versions, shall we use more fine-graded key name? e.g. obs-topology?

Comment (YianZong) [reply operatorconfig/moduleconfig/common/version-values.yaml:13]: @forrestxia this is observability module version 1.3.0, rather than individual component image version

**Comment by YianZong** [comment]
> Hi @YianZong Please do not merge this PR as posting activities are in progress for CSM Operator. Once those are done, we can merge. Blocking this PR for now.

Hi @rajendraindukuri, when will the posting be done?

**Comment by rajendraindukuri** [comment]
> > Hi @YianZong Please do not merge this PR as posting activities are in progress for CSM Operator. Once those are done, we can merge. Blocking this PR for now.
> 
> Hi @rajendraindukuri, when will the posting be done?

> Hi @YianZong Since RTS was done yesterday. We started posting today.. If all goes good, should be done within a week. Will keep you posted

File: operatorconfig/moduleconfig/common/version-values.yaml
Comment (forrestxia) [reply operatorconfig/moduleconfig/common/version-values.yaml:13]: observability module version is an abstract concept-only version. If so, shall we use v1.4.0 for the coming release? @YianZong

Comment (YianZong) [reply operatorconfig/moduleconfig/common/version-values.yaml:13]: @forrestxia We add v1.3.0 first, then add v1.4.0 for Q4 release.

Comment (forrestxia) [reply operatorconfig/moduleconfig/common/version-values.yaml:13]: ok, see. thanks for explanation.

File: pkg/modules/observability.go
Comment (taohe1012) [inline pkg/modules/observability.go:108]: this line could be deleted

File: controllers/csm_controller.go
Comment (taohe1012) [inline controllers/csm_controller.go:646]: for this block codes, how about Extract them into a function

Comment (taohe1012) [reply controllers/csm_controller.go:646]: because in the future, metrics codes will be added to this block too

File: pkg/modules/observability.go
Comment (taohe1012) [inline pkg/modules/observability.go:49]: better rename `topo csmv1.Module` to `obs csmv1.Module`

File: samples/storage_csm_powerscale_v240.yaml
Comment (YianZong) [reply samples/storage_csm_powerscale_v240.yaml:311]: Observability v1.3.0 aligns with csi-powerscale 2.4

File: pkg/modules/observability.go
Comment (YianZong) [reply pkg/modules/observability.go]: addressed

File: controllers/csm_controller.go
Comment (YianZong) [reply controllers/csm_controller.go:641]: addressed

File: pkg/modules/observability.go
Comment (YianZong) [reply pkg/modules/observability.go]: addressed

**Comment by YianZong** [comment]
Hi Code Owners, please kindly review the PR and let me know if any change needed. Thank you!

**Comment by forrestxia** [comment]
@YianZong "Check for forbidden words" seems required, and no result yet, please check what to be done.

**Comment by YianZong** [comment]
> @YianZong "Check for forbidden words" seems required, and no result yet, please check what to be done.

@forrestxia The check of "Check for forbidden words" is commented as removed in the actions.yml.

**Comment by YianZong** [comment]
Rebased main branch

**Review by taohe1012 — APPROVED** [review]

**Review by baoy1 — DISMISSED** [review]

File: api/v1alpha1/types.go
Comment (rensyct) [inline api/v1alpha1/types.go:34]: spell check Oberservability -> Observability

File: api/v1alpha1/types.go
Comment (rensyct) [inline api/v1alpha1/types.go:33]: same comment as below

File: api/v1alpha1/types.go
Comment (rensyct) [inline api/v1alpha1/types.go:57]: Same comment as above

File: controllers/csm_controller.go
Comment (rensyct) [inline controllers/csm_controller.go:511]: spell check Deleteing -> deleting

File: controllers/csm_controller.go
Comment (rensyct) [inline controllers/csm_controller.go:826]: Same comment as above

File: pkg/modules/testdata/cr_powerscale_observability.yaml
Comment (rensyct) [inline pkg/modules/testdata/cr_powerscale_observability.yaml:9]: Should this not be 2.4.0?

File: pkg/modules/testdata/cr_powerscale_observability.yaml
Comment (rensyct) [inline pkg/modules/testdata/cr_powerscale_observability.yaml:13]: Same comment as above

**Review by rensyct — DISMISSED** [review]

**Comment by bharathsreekanth** [comment]
@YianZong v1alpha1 to v1 changes will affect some of the code in here. I personally preferred to have that change go in first, so that all the teams can do the changes needed based on that. Let us discuss this in the meeting today to see how to coordinate.

**Comment by YianZong** [comment]
Thanks, @rensyct all addressed.

**Comment by YianZong** [comment]
Rebased main branch and updated v1alpha to v1 in newly added files.
Manually verified again:
image

**Review by forrestxia — APPROVED** [review]

**Review by rensyct — APPROVED** [review]

**Review by taohe1012 — APPROVED** [review]

**Pull request merged**
